rclpy/rclpy/impl/rcutils_logger.py

```diff
@@ -0,0 +1,310 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from collections import OrderedDict
+import importlib
+import inspect
+from pathlib import Path
+import time
+
+_rclpy_logging = importlib.import_module('._rclpy_logging', package='rclpy')
+
+# Known filenames from which logging methods can be called (will be ignored in `_find_caller`).
+_internal_callers = []
+# This will cause rclpy filenames to be registered in `_internal_callers` on first logging call.
+_populate_internal_callers = True
+
+
+def _normalize_path(path):
+    return str(Path(path).resolve())
+
+
+def _find_caller(frame):
+    """Get the first calling frame that is outside of rclpy."""
+    global _populate_internal_callers
+    global _internal_callers
+    if _populate_internal_callers:
+        # Populate the list of internal filenames from which logging methods can be called.
+        # This has to be done from within a function to avoid cyclic module imports.
+        import rclpy.logging
+        # Preserve any filenames that may have been added by third parties.
+        _internal_callers.extend([
+            _normalize_path(__file__),
+            _normalize_path(rclpy.logging.__file__),
+        ])
+        _populate_internal_callers = False
+
+    frame_filename = _normalize_path(inspect.getabsfile(frame))
+    while any(f in frame_filename for f in _internal_callers):
+        frame = frame.f_back
+        frame_filename = _normalize_path(inspect.getabsfile(frame))
+    return frame
+
+
+class CallerId:
```

Comment:
Maybe the class could use `__slots__` if it is never intended to be extended with additional attributes.

That would also make the implementation of `__hash__` and `__eq__` easier by just iterating the slots and avoiding the repetition of the member names.

Reply:
I made use of `__slots__` as you suggested in d34a86cce78a4623cceceda5d66394d760ba121b but then figured what we really want is a named tuple so switched to that in b4a87876cbd5a9d031cb0e280f22cd4369d134d5

Reply:
Even nicer! :1st_place_medal: